### DISCUSSION TITLE
Cache coherency issue

### DISCUSSION THREAD

**Discussion by AntonTimofeev** [gem5/gem5#1336] [gem5-dev]
Hi all,

I've faced a problem with cache coherency and looking for any advice on how to handle it.

I have two cores with private L1, L2 and shared L3. These cores share a single cacheline.
L1DCache_0 has it as shared.
L1DCache_1 owns it.

![image](https://github.com/gem5/gem5/assets/7997841/8878aa61-7e4a-4523-b709-1f6256385653)

L1DCache_0 creates UpgradeReq and waits for answer
L1DCache_1 gives the permission via UpgradeResp, and shortly after that issues CleanInvalidateReq (CPU1 executes a `clflush` instruction).

While UpgradeResp is travelling through the cache hierarchy as a TimingResp, CleanInvalidate overtakes it as a SnoopReq, gets to L1DCache_0 cacheline, wipes it, and when UpgradeResp arrives there is no such cacheline already.

Thus, I got an assertion in `handleFill()` method. https://github.com/gem5/gem5/blob/stable/src/mem/cache/base.cc#L1538

\
Thanks in advance!
Anton

**Comment by powerjg** [comment]
This doesn't answer your question, but I would suggest using Ruby (maybe the CHI protocol with your own custom hierarchy) instead of classic caches if you're doing multicore/coherence-based workloads.

Slightly more to your question, this is likely a bug. I also wonder if the L3XBar is a "point of coherence" and not converting the UpgradeResp correctly? It's hard to know. The classic cache coherence protocol is pretty brittle.

**Reply by AntonTimofeev** [reply]
Probably switching to Ruby could be the answer, thanks.

Yes, L3XBar is set as a `point_of_coherency`.
Could you please elaborate on "converting correctly"? Right now it receives `UpgradeResp` via `recvTimingSnoopResp` and routes it to `L2Cache_0` using `sendTiming`, so `L2Cache_0` gets it via `recvTimingResp`.

What could be the correct behavior in this case:
- defer `CleanInvalidReq` until `UpgradeResp` is arrived
- treat `UpgradeResp` as an error response if cache block could not be found
- something else
?

**Reply by powerjg** [reply]
I have no specific ideas on how to fix the problem 🙂 

Without digging into the details of the classic protocol, I would have expected that the point of coherence would convert the UpgradeResp to a normal data response (via accessing memory since it accompanies a CleanIvalidateReq?). But that's just a gut feeling based on general coherence knowledge. 

The classic cache's coherence protocol is brittle. There are definitely race conditions that aren't handled, and many of the race conditions that are handled have so many special case clauses that they are likely to be wrong. It's been tweaked a lot in the ~15-20 years that it's been around, and that's caused significant bitrot and increased brittleness (not to mention that it was not designed for x86 or highly parallel systems). 

So, there's likely a solution to fix the problem you ran into. It may be a small change somewhere (my gut says at the point of coherency, but I could be wrong). Or, it could be a much bigger update to the overall protocol to handle some new race condition. If you can figure out the former, we would welcome the fix! However, I would advise just moving to ruby (again likely with a specialized CHI hierarchy) to as the most straightforward solution.
